fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/RelativePriorityPlanSelector.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * RelativePriority.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2021 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.query.plan.plans;
+
+import com.apple.foundationdb.record.ObjectPlanHash;
+import com.apple.foundationdb.record.PlanHashable;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Verify;
+import com.google.common.collect.ImmutableList;
+
+import javax.annotation.Nonnull;
+import java.util.List;
+import java.util.Objects;
+import java.util.Random;
+
+public class RelativePriorityPlanSelector implements RecordQuerySelectorPlan.PlanSelector {
+    private static final ObjectPlanHash BASE_HASH = new ObjectPlanHash("Relative-Priority-Plan-Selector");
+    @Nonnull
+    List<Double> priorities;
+    @Nonnull
+    Random random;
+
+    public RelativePriorityPlanSelector(@Nonnull final List<Double> priorities) {
+        this(priorities, new Random());
+    }
+
+    @VisibleForTesting
+    RelativePriorityPlanSelector(@Nonnull final List<Double> priorities, @Nonnull final Random random) {
+        Verify.verify(!priorities.isEmpty());
+        this.priorities = ImmutableList.copyOf(priorities);
+        this.random = random;
```

Comment:
`Random` performance is pretty bad especially if there is contention between threads. Consider using `ThreadLocalRandom` here even though it forsakes stability in multithreaded environments.

Reply:
In theory, if we cared about using a `ThreadLocalRandom` in multiple threads, we could make the `random` field of the class `@Nullable`, and then use `ThreadLocalRandom.current()` if `this.random == null`

Reply:
Done. random.nextInt() is called once per query (since it will use continuation value after that). Using ThreadLocalRandom.current.nextInt() initializes it once per thread, which is the requests thread and so the same random instance will be used for all plans for this query.